File: python/algo/leecode/algo_lcp10.py
# Definition for a binary tree node.
import sys
sys.path.append(sys.path[0] + "/../..")
import typing
from common import TreeNode, createTreeNode

def dfs(root: TreeNode) -> typing.Tuple[int,float]:
    if root is None: return (0, 0.0)
    lf, ls = dfs(root.left)
    rf, rs = dfs(root.right)

    return  (lf + rf + root.val, root.val + max(max(ls, rs), (lf + rf) / 2))


class Solution:

    def minimalExecTime(self, root: TreeNode) -> float:
        f,s = dfs(root)
        return s

    def minimalExecTime2(self, root):
        """
        :type root: TreeNode
        :rtype: float
        """
        def dfs(root):
            if root is None:
                return 0., 0.
            tc = root.val
			
            # 先递归左右子树
            a, b = dfs(root.left)
            c, d = dfs(root.right)
            
            tc = tc + a + c
            # 只考虑 a >= c 的情况，不满足就交换
            if a < c:
                a, c = c, a
                b, d = d, b
            
            if a - 2 * b <= c:
                pc = (a + c) / 2
            else:
                pc = c + b

            return tc, pc

        tc, pc = dfs(root)
        return tc - pc


if __name__ == "__main__":
    sol = Solution()
    # nums = [7,5,6,4]
    l1 = [1,3,3,None,None,4,4]

    root = createTreeNode(l1)
    print(sol.minimalExecTime2(root))

File: cplus/leetcode/202508/algolcp10.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 二叉树 - 2.12 树形 DP - 复习
class Solution {
public:
    double minimalExecTime(TreeNode* root) {
        // 参考题解 - 解法二
        // return pair<double, double> 分别为全部执行时间，最小执行时间（并行）
        auto dfs = [&](this auto&& dfs, TreeNode* node)->pair<double, double>{
            if (node == nullptr){
                return {0, 0};
            }

            auto [l_tot_time, l_min_time] = dfs(node->left);
            auto [r_tot_time, r_min_time] = dfs(node->right);
            return {l_tot_time + r_tot_time + node->val, 
                node->val + max((l_tot_time+r_tot_time)/2.0, max(l_min_time, r_min_time))};
        };
        return dfs(root).second;
    }

    double minimalExecTime2(TreeNode* root) {
        // 参考题解 - 解法一
        // return pair<double, double> 分别为全部执行时间，并行时间，（最小时间=全部时间-并行时间）
        auto dfs = [&](this auto&& dfs, TreeNode* node)->pair<double, double>{
            if (node == nullptr){
                return {0, 0};
            }

            auto [a, b] = dfs(node->left);
            auto [c, d] = dfs(node->right);
            int a, b, c, c;     // a >= c
            if (a < c) {
                swap(a, c); swap(b, d);
            } 

            // a-2*b 为串行时间，正好可以和 c 并行执行
            if (a-2*b<=c) return {a+c+node->val, (a+c)/2.0};
            else return {a+c+node->val, b+c};

        };
        auto ans = dfs(root);
        return ans.first - ans.second;
    }    
};

